Confirm the answer according to #5 (#6)
tc39-transfer/proposal-object-pick-or-omit: README.md

    - square brackets:

        ```js
        ({a : 1, b : 2, c : 3}).['a', 'b']; // => {a: 1, b: 2}

        const keys = ['a', 'b'];
        ({a : 1, b : 2, c : 3}).[keys[0], keys[1]]; // => {a: 1, b: 2}
        ({a : 1, b : 2, c : 3}).[...keys]; // => {a: 1, b: 2}
        ```

    - curly brackets

        ```js
        ({a : 1, b : 2, c : 3}).{a, b} // => {a: 1, b: 2}

        const keys = ['a', 'b'];
        ({a : 1, b : 2, c : 3}).{[keys[0]], b}; // => {a: 1}
        ({a : 1, b : 2, c : 3}).{[...keys]}; // => {a: 1, b: 2}

        // Similar to destructuring
        ({a : 1, b : 2, c : 3}).{a, b : B}; // => {a: 1, B: 2}
        ```

>         Currently, there is a disagreement on whether properties with default assignment values should be picked.

Reviewer: what disagreement exactly?

Author: Exactly on the following two cases:

```js
// not picked
({a : 1, b : 2, c : 3}).{a, d = 2}; // => {a : 1}

// picked
({a : 1, b : 2, c : 3}).{a, d = 2}; // => {a : 1, d : 2}
```

Reviewer: Obviously it would have to be picked - the `.{ … }` denotes what the new object should contain. Whether the original object has a `d` key that's `undefined` versus lacks one is irrelevant, since almost every other part of the language doesn't differentiate here, including destructuring.

What's the argument for "not picked"?

Author: The argument is when considering the meaning of picking: we can't pick off a non-existed property from an object.

It is similar to:

```js
Object.pick({a : 1, b : 2, c : 3}, ['a', 'd']); // => {a : 1}
```

Reviewer: Sure you can - just like you can Get a nonexistent property.

You can also do `Object.pick({}, ['toString'])` and obviously you'd get `{ toString: Object.prototype.toString }` - why would an absent property be any different?

Author: Of course, the "absent" means for owned properties or proto properties at the same time.

Author: I have confirmed it in the document.

        ({a : 1, b : 2, c : 3}).{a, d = 2}; // => {a : 1, d : 2}
        ```

    Nevertheless, it is just a simple vision, and feel free to discuss it.

### FAQ

1. When it comes to the prototype chain of an object, should the method pick or omit it?

>     A: Consistent with destructuring: we can pick off properties of the prototype, but can't omit them.

Reviewer: ```suggestion
    A: Consistent with destructuring: we can explicitly pick off properties of the prototype, but can't avoid omitting them.
```

Author: but avoid omitting them?

Reviewer: it can not avoid omitting them - all prototype properties are ignored.

Author: There is an ambiguity between ommiting them by calling `Object.pick` or `Object.omit`. Just say all not owned properties are ignored by calling `Object.omit`?

Reviewer: I think Object.prototype is a special case.

Just like object spread/rest syntax and destructuring, `Object.pick` and `Object.omit` will Get explicitly named properties (which includes own or inherited), but will not omit inherited properties, and the final object will inherit from Object.prototype, not whatever the input inherits from.

    ```js
    Object.pick({a : 1}, ['toString']); // => {toString: f}
    // equivalent to the behavior
    const {toString} = {a : 1};
    toString; // => f toString() { [native code] }

    Object.omit({a : 1}, ['toString']).toString; // => ƒ toString() { [native code] }
    // equivalent to the behavior
    const {toString, ...res} = {a : 1}

>     res.toString; // => ƒ toString() { [native code] }

Reviewer: no - toString can’t ever show up since it’s not an own property. Things in the toOmit array are ignored if they aren’t an own property.

Author: Only describe like this?

```js
// avoid omitting properties which are not owned
Object.omit({a : 1}, ['toString']).toString; // => ƒ toString() { [native code] }
```

Reviewer 2: `toString` is present on the result because the result is an instance of `Object`, and instances of `Object` has `toString` method. It has nothing to do with the fact that `toString` is mentioned in the second argument array except to represent that even though it's been explicitly mentioned it would still be present on the result

Although, I agree that we can tell the same in a less sophisticated way just by saying that non-own properties in the second argument array for `Object.omit` are ignored

Author: Right, it is not a good case because `Object.omit` should return a plain object, so it seems it is meanless to discuss about whether it will omit the properties which an object has not owned. The method does not modify the original object at all.

Reviewer: It's probably better to show an example of something that inherits from "not Object" - like, perhaps, an array trying to omit `filter`.

    ```

    The implementation of [`_.pick`](https://lodash.com/docs/4.17.15#pick) and [`_.omit`](https://lodash.com/docs/4.17.15#omit) by Lodash has also taken care about the chain.

    The same rule applies to `__proto__` event if it has been [deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto) because the proposal should be pure enough to not specify a special logic to eliminate deprecated properties:

    ```js
    Object.pick({}, ['__proto__']); // => {__proto__: {...}}
    // equivalent to the behavior
    const {__proto__} = {};

>     __proto__; // => {...}

Reviewer: ```suggestion
    __proto__; // => Object.prototype
```

>     Object.omit({}, ['__proto__']).__proto__; // => {...}

Reviewer: ```suggestion
    Object.omit({}, ['__proto__']).__proto__; // => Object.prototype
```

    // equivalent to the behavior
    const {__proto__, ...res} = {}

>     res.__proto__; // => {...}

Reviewer: ```suggestion
    res.__proto__; // => Object.prototype
```

    const omitOwn = (obj, keys) => Object.omit(obj, keys.filter(key => obj.hasOwnProperty(key)));
    ```

2. What is the type of returned value?

    A: Consistent with destructuring: should return plain objects.

    ```js
    Object.pick([]); // => {}
    Object.omit([]); // => {}
    Object.pick(new Map()); // => {}
    Object.omit(new Map()); // => {}
    
    // equivalent to the behavior
    const {...res} = [];
    res instanceof Array; // => false
    const {...res} = new Map();
    res instanceof Map; // => false
    const {...res} = new Set();
    res instanceof Set; // => false
    ```

3. How to handle `Symbol`?

    A: Consistent with destructuring.

    ```js
    Object.pick([], [Symbol.iterator]); // => {Symbol(Symbol.iterator): f}, pick off from the prototype
    // equivalent to the behavior
    const {[Symbol.iterator] : iter} = [];

>     iter; // => Symbol(Symbol.iterator): f

Reviewer: ```suggestion
    iter; // => Array.prototype[Symbol.iterator)]
```

    
    Object.omit([], [Symbol.iterator]); // => {}, plain objects
    // equivalent to the behavior
    const {[Symbol.iterator] : iter, ...res} = [];
    res instanceof Array; // => false
    
    const symbol = Symbol('key');
    Object.omit({a : 1, [symbol]: 2}, [symbol]); // => {a : 1}
    // equivalent to the behavior
    const {[symbol] : _, ...res} = {a : 1, [symbol]: 2};
    res; // => {a : 1}
    
    Object.prototype[symbol] = 'test'; // override prototype
    Object.pick({}, [symbol]); // => {Symbol(key): "test"}, pick off from the prototype
    Object.omit({}, [symbol])[symbol]; // => "test", cannot omit properties from the prototype
    // equivalent to the behavior
    const {[symbol] : sym, ...res} = {};

>     sym; // => 'test'

Reviewer: ```suggestion
    sym; // => symbol
```

Author: The `sym` should be 'test' rather than the symbol.

    ```js
    postData({[key1] : o[key1], [key2] : o[key2]});
    postData(Object.pick(o, [key1, key2]));
    ```

6. Why can't be defined on the `Object.prototype` directly?

>     A: As `Object` is especially fundamental, both of them will result in conflicts of properties of any other objects. In shorthand, if defined, any objects inherited from `Object` with `pick` or `omit` defined in its prototype should break.

Reviewer: ```suggestion
    A: As `Object` is especially fundamental, both of them will result in conflicts of properties of any other objects. In shorthand, if defined, any objects inherited from `Object` with `pick` or `omit` defined in its prototype would break. Also, objects that inherit from `null` would be left unable to use this functionality.
```
